File: src/core/minimize_energy.cpp
/*
  Copyright (C) 2010-2018 The ESPResSo project
  Copyright (C) 2002,2003,2004,2005,2006,2007,2008,2009,2010
  Max-Planck-Institute for Polymer Research, Theory Group

  This file is part of ESPResSo.

  ESPResSo is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  ESPResSo is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "minimize_energy.hpp"
#include "cells.hpp"
#include "communication.hpp"
#include "event.hpp"
#include "integrate.hpp"
#include "rotation.hpp"

#include <utils/math/sqr.hpp>

#include <boost/mpi/collectives/all_reduce.hpp>
#include <boost/mpi/operations.hpp>

#include <algorithm>
#include <limits>

#ifdef MINIMIZE_ENERGY_DEBUG
#define MINIMIZE_ENERGY_TRACE(A) A
#else
#define MINIMIZE_ENERGY_TRACE(A)
#endif

struct MinimizeEnergyParameters {
  double f_max;
  double gamma;
  int max_steps;
  double max_displacement;
};

static MinimizeEnergyParameters *params = nullptr;

/* Signum of val */
template <typename T> int sgn(T val) { return (T(0) < val) - (val < T(0)); }

bool steepest_descent_step() {
  // Maximal force encountered on node
  double f_max = -std::numeric_limits<double>::max();
  // and globally

  // Positional increments
  double dp, dp2, dp2_max = -std::numeric_limits<double>::max();

  // Iteration over all local particles

  for (auto &p : local_cells.particles()) {
    auto f = 0.0;

    dp2 = 0.0;
    // For all Cartesian coordinates
    for (int j = 0; j < 3; j++) {
#ifdef EXTERNAL_FORCES
      // Skip, if coordinate is fixed
      if (!(p.p.ext_flag & COORD_FIXED(j)))
#endif
#ifdef VIRTUAL_SITES
        // Skip positional increments of virtual particles
        if (!p.p.is_virtual)
#endif
        {
          // Square of force on particle
          f += Utils::sqr(p.f.f[j]);

          // Positional increment
          dp = params->gamma * p.f.f[j];
          if (fabs(dp) > params->max_displacement)
            // Crop to maximum allowed by user
            dp = sgn<double>(dp) * params->max_displacement;
          dp2 += Utils::sqr(dp);

          // Move particle
          p.r.p[j] += dp;
          MINIMIZE_ENERGY_TRACE(printf("part %d dim %d dp %e gamma*f %e\n", i,
                                       j, dp, params->gamma * p.f.f[j]));
        }
    }
#ifdef ROTATION
    {
      // Rotational increment
      auto const dq = params->gamma * p.f.torque; // Vector parallel to torque
      auto const t = p.f.torque.norm2();

      // Normalize rotation axis and compute amount of rotation
      auto const l = dq.norm();
      if (l > 0.0) {
        auto const axis = dq / l;
        auto const angle = (std::abs(l) > params->max_displacement)
                               ? sgn(l) * params->max_displacement
                               : l;

        // Rotate the particle around axis dq by amount l
        local_rotate_particle(p, axis, angle);
      }

      f_max = std::max(f_max, t);
    }
#endif
    // Note maximum force/torque encountered
    f_max = std::max(f_max, f);
    dp2_max = std::max(dp2_max, dp2);
  }

  set_resort_particles(Cells::RESORT_LOCAL);

  // Synchronize maximum force/torque encountered
  namespace mpi = boost::mpi;
  auto const f_max_global =
      mpi::all_reduce(comm_cart, f_max, mpi::maximum<double>());

  // Return true, if the maximum force/torque encountered is below the user
  // limit.
  return (sqrt(f_max_global) < params->f_max);
}

void minimize_energy_init(const double f_max, const double gamma,
                          const int max_steps, const double max_displacement) {
  if (!params)
    params = new MinimizeEnergyParameters;

  params->f_max = f_max;
  params->gamma = gamma;
  params->max_steps = max_steps;
  params->max_displacement = max_displacement;
}

void minimize_energy() {
  if (!params)
    params = new MinimizeEnergyParameters;

  MPI_Bcast(params, sizeof(MinimizeEnergyParameters), MPI_BYTE, 0, comm_cart);
  int integ_switch_old = integ_switch;
  integ_switch = INTEG_METHOD_STEEPEST_DESCENT;
  integrate_vv(params->max_steps, -1);
  integ_switch = integ_switch_old;
}

File: testsuite/python/minimize_energy.py
# Copyright (C) 2010-2019 The ESPResSo project
#
# This file is part of ESPResSo.
#
# ESPResSo is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# ESPResSo is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
import unittest as ut
import unittest_decorators as utx
import numpy as np

import espressomd


@utx.skipIfMissingFeatures("LENNARD_JONES")
class test_minimize_energy(ut.TestCase):

    np.random.seed(42)
    system = espressomd.System(box_l=[10.0, 10.0, 10.0])

    test_rotation = espressomd.has_features(("ROTATION", "DIPOLES"))
    if test_rotation:
        from espressomd.constraints import HomogeneousMagneticField

    box_l = 10.0
    density = 0.6
    vol = box_l**3
    n_part = int(vol * density)

    lj_eps = 1.0
    lj_sig = 1.0
    lj_cut = 1.12246

    def setUp(self):
        self.system.box_l = 3 * [self.box_l]
        self.system.cell_system.skin = 0.4
        self.system.time_step = 0.01
        self.system.non_bonded_inter[0, 0].lennard_jones.set_params(
            epsilon=self.lj_eps, sigma=self.lj_sig,
            cutoff=self.lj_cut, shift="auto")
        if self.test_rotation:
            self.system.constraints.add(
                self.HomogeneousMagneticField(H=[-0.5, 0, 0]))

    def tearDown(self):
        self.system.part.clear()
        self.system.integrator.set_vv()

    def test_relaxation(self):
        for i in range(self.n_part):
            p = self.system.part.add(
                id=i, pos=np.random.random(3) * self.system.box_l)
            if self.test_rotation:
                p.dip = np.random.random(3)
                p.dipm = 1
                p.rotation = (1, 1, 1)

        self.assertNotAlmostEqual(
            self.system.analysis.energy()["total"], 0, places=10)

        self.system.integrator.set_steepest_descent(
            f_max=0.0, gamma=0.1, max_displacement=0.05)

        self.system.integrator.run(500)

        self.system.constraints.clear()

        # Check
        self.assertAlmostEqual(
            self.system.analysis.energy()["total"], 0, places=10)
        np.testing.assert_allclose(np.copy(self.system.part[:].f), 0.)
        if self.test_rotation:
            np.testing.assert_allclose(np.copy(self.system.part[:].dip),
                                       np.hstack((-np.ones((self.n_part, 1)), np.zeros((self.n_part, 1)), np.zeros((self.n_part, 1)))), atol=1E-9)

    def test_rescaling(self):
        self.system.part.add(pos=[5., 5., 4.9], type=0)
        self.system.part.add(pos=[5., 5., 5.1], type=0)

        self.system.non_bonded_inter[0, 0].lennard_jones.set_params(
            epsilon=self.lj_eps, sigma=self.lj_sig,
            cutoff=self.lj_cut, shift="auto")

        self.system.integrator.run(0)
        f_old = np.copy(self.system.part[:].f)

        # No-op, because gamma = 0.
        self.system.integrator.set_steepest_descent(
            f_max=0.0, gamma=0.0, max_displacement=0.001)
        self.system.integrator.run(1)

        np.testing.assert_allclose(f_old, np.copy(self.system.part[:].f))


if __name__ == "__main__":
    ut.main()
